addons/website_sale/static/src/js/website_sale_service.js

```diff
@@ -0,0 +1,541 @@
+            const selectedCustomPtav = ptal.attribute_values.find(
+                ptav => ptav.is_custom && selectedPtavIds.has(ptav.id)
+            );
+            if (selectedCustomPtav) {
+                serializedProduct.product_custom_attribute_values.push({
+                    custom_product_template_attribute_value_id: selectedCustomPtav.id,
+                    custom_value: ptal.customValue ?? '',
+                });
+            }
+        }
+
+        // No variant attributes.
+        serializedProduct.no_variant_attribute_value_ids = product.attribute_lines
+            .filter(ptal => ptal.create_variant === 'no_variant')
+            .flatMap(ptal => ptal.selected_attribute_value_ids);
+
+        return serializedProduct;
+    }
+
+    /**
+     * Serialize a combo item into a format understandable by the server.
+     *
+     * @private
+     * @param {ProductComboItem} comboItem - The combo item to serialize.
+     * @param {Number} parentProductTemplateId - The parent's product template id, as a
+     *      `product.template` id.
+     * @param {Number} quantity - The quantity to add to the cart.
+     *
+     * @returns {Object} - The serialized combo item.
+     */
+    _serializeComboItem(comboItem, parentProductTemplateId, quantity) {
+        return {
+            product_template_id: comboItem.product.product_tmpl_id,
+            parent_product_template_id: parentProductTemplateId,
+            quantity: quantity,
+            ...serializeComboItem(comboItem),
+        };
+    }
+
+    //--------------------------------------------------------------------------
+    // Helpers
+    //--------------------------------------------------------------------------
+
+    /**
+     * Asynchronously adds a product to the shopping cart.
+     *
+     * @async
+     * @private
+     * @param {Object} addToCartData - Data containing product(s) to add to the cart and options for
+     *      adding them.
+     * @param {Number} addToCartData.productTemplateId - The product template's id, as a
+     *      `product.template` id.
+     * @param {Number} addToCartData.productId - The product's id, as a `product.product` id.
+     * @param {Number} addToCartData.quantity - The quantity of the product to add to the cart.
+     * @param {CustomAttributeValues[]} [addToCartData.productCustomAttributeValues=[]] - An
+     *      array of objects representing custom attribute values for the product.
+     * @param {Number[]} [addToCartData.noVariantAttributeValues=[]] - The selected non-stored
+     *      attribute(s), as a list of `product.template.attribute.value` ids.
+     * @param {Boolean} [addToCartData.shouldRedirectToCart=false] - Whether to redirect the
+     *      customer to the cart. Defaults to false.
+     * @param {...*} [addToCartData.rest] - Locally unused data sent to the controllers.
+     *
+     * @returns {Number} - The product's quantity added to the cart.
+     */
+    async _addToCart({
```

Comment:
If it has a different purpose it should have a different name.

Reply:
IMO it is the right name, it's the method that calls the server to effectively add a product to the cart, whereas the public one checks if the product is well configured before calling this private one.

Reply:
Maybe it's the other that has the wrong name, but there is definitely a bad design if two homonym methods perform different jobs.